analyzer/src/main/kotlin/managers/SPM.kt

```diff
@@ -0,0 +1,277 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.analyzer.managers
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+
+import java.io.File
+import java.net.URI
+
+import org.ossreviewtoolkit.analyzer.AbstractPackageManagerFactory
+import org.ossreviewtoolkit.analyzer.PackageManager
+import org.ossreviewtoolkit.analyzer.PackageManagerResult
+import org.ossreviewtoolkit.analyzer.managers.utils.SPMDependencyHandler
+import org.ossreviewtoolkit.downloader.VcsHost
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.model.DependencyGraph
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Project
+import org.ossreviewtoolkit.model.ProjectAnalyzerResult
+import org.ossreviewtoolkit.model.RemoteArtifact.Companion.EMPTY
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
+import org.ossreviewtoolkit.model.config.RepositoryConfiguration
+import org.ossreviewtoolkit.model.orEmpty
+import org.ossreviewtoolkit.model.utils.DependencyGraphBuilder
+import org.ossreviewtoolkit.utils.common.ProcessCapture
+import org.ossreviewtoolkit.utils.core.log
+import org.ossreviewtoolkit.utils.core.normalizeVcsUrl
+
+interface SPMCLIExecutor {
+    fun executeSwift(definitionFile: File): JsonNode
+}
+
+/**
+ * The [Swift Package Manager](https://github.com/apple/swift-package-manager).
+ */
+class SPM(
+    name: String,
+    analysisRoot: File,
+    analyzerConfig: AnalyzerConfiguration,
+    repoConfig: RepositoryConfiguration,
+    private val cliExecutor: SPMCLIExecutor
+) : PackageManager(name, analysisRoot, analyzerConfig, repoConfig) {
+
+    private val graphBuilder = DependencyGraphBuilder(SPMDependencyHandler())
+
+    class Factory : AbstractPackageManagerFactory<SPM>(MANAGER_NAME), SPMCLIExecutor {
+        override val globsForDefinitionFiles = listOf("Package.*", ".package.resolved")
+
+        override fun executeSwift(definitionFile: File): JsonNode {
+            val process = ProcessCapture(
+                workingDir = definitionFile.parentFile,
+                command = arrayOf("swift", "package", "show-dependencies", "--format", "json")
+            ).requireSuccess()
+            return ObjectMapper().readTree(process.stdout)
```

Comment:
Unfortunately, this only works if the project was build once before.
`swift package show-dependencies --format json` prints "debug" output also to `stdout` which then cannot be parsed as a JSON.

To reproduce, you have to remove all files not under version control: `git clean -fdx`.
And then run `swift package show-dependencies --format json`:

This is the output I get:
<details>
    <summary>swift output log</summary>

```
~/workspace/playground/vapor main
❯ swift package show-dependencies --format json
Fetching https://github.com/vapor/websocket-kit.git from cache
Fetching https://github.com/vapor/multipart-kit.git from cache
Fetching https://github.com/apple/swift-metrics.git from cache
Fetched https://github.com/vapor/multipart-kit.git (0.48s)
Fetching https://github.com/apple/swift-log.git from cache
Fetched https://github.com/vapor/websocket-kit.git (0.48s)
Fetching https://github.com/apple/swift-nio-extras.git from cache
Fetched https://github.com/apple/swift-metrics.git (0.49s)
Fetching https://github.com/apple/swift-nio-http2.git from cache
Fetched https://github.com/apple/swift-log.git (0.49s)
Fetching https://github.com/apple/swift-nio-ssl.git from cache
Fetched https://github.com/apple/swift-nio-extras.git (0.49s)
Fetching https://github.com/apple/swift-nio.git from cache
Fetched https://github.com/apple/swift-nio-http2.git (0.49s)
Fetching https://github.com/swift-server/swift-backtrace.git from cache
Fetched https://github.com/swift-server/swift-backtrace.git (0.46s)
Fetching https://github.com/vapor/routing-kit.git from cache
Fetched https://github.com/apple/swift-nio-ssl.git (0.49s)
Fetching https://github.com/apple/swift-crypto.git from cache
Fetched https://github.com/apple/swift-nio.git (0.76s)
Fetching https://github.com/vapor/console-kit.git from cache
Fetched https://github.com/vapor/routing-kit.git (0.44s)
Fetching https://github.com/vapor/async-kit.git from cache
Fetched https://github.com/apple/swift-crypto.git (0.43s)
Fetching https://github.com/swift-server/async-http-client.git from cache
Fetched https://github.com/vapor/console-kit.git (0.46s)
Fetched https://github.com/vapor/async-kit.git (0.45s)
Fetched https://github.com/swift-server/async-http-client.git (0.52s)
Computing version for https://github.com/apple/swift-nio-http2.git
Computed https://github.com/apple/swift-nio-http2.git at 1.20.1 (0.43s)
Computing version for https://github.com/swift-server/async-http-client.git
Computed https://github.com/swift-server/async-http-client.git at 1.9.0 (0.27s)
Fetching https://github.com/apple/swift-nio-transport-services.git from cache
Fetched https://github.com/apple/swift-nio-transport-services.git (0.46s)
Computing version for https://github.com/apple/swift-nio-transport-services.git
Computed https://github.com/apple/swift-nio-transport-services.git at 1.11.4 (0.27s)
Computing version for https://github.com/apple/swift-nio.git
Computed https://github.com/apple/swift-nio.git at 2.39.0 (0.28s)
Computing version for https://github.com/apple/swift-log.git
Computed https://github.com/apple/swift-log.git at 1.4.2 (0.25s)
Computing version for https://github.com/apple/swift-nio-extras.git
Computed https://github.com/apple/swift-nio-extras.git at 1.10.2 (0.26s)
Computing version for https://github.com/vapor/multipart-kit.git
Computed https://github.com/vapor/multipart-kit.git at 4.5.1 (0.25s)
Computing version for https://github.com/apple/swift-metrics.git
Computed https://github.com/apple/swift-metrics.git at 2.3.0 (0.26s)
Computing version for https://github.com/swift-server/swift-backtrace.git
Computed https://github.com/swift-server/swift-backtrace.git at 1.3.1 (0.26s)
Computing version for https://github.com/vapor/routing-kit.git
Computed https://github.com/vapor/routing-kit.git at 4.3.1 (0.26s)
Computing version for https://github.com/vapor/websocket-kit.git
Computed https://github.com/vapor/websocket-kit.git at 2.3.0 (0.26s)
Computing version for https://github.com/apple/swift-nio-ssl.git
Computed https://github.com/apple/swift-nio-ssl.git at 2.18.0 (0.29s)
Computing version for https://github.com/vapor/console-kit.git
Computed https://github.com/vapor/console-kit.git at 4.2.7 (0.27s)
Computing version for https://github.com/apple/swift-crypto.git
Computed https://github.com/apple/swift-crypto.git at 2.0.5 (0.27s)
Computing version for https://github.com/vapor/async-kit.git
Computed https://github.com/vapor/async-kit.git at 1.11.1 (0.27s)
Creating working copy for https://github.com/vapor/console-kit.git
Working copy of https://github.com/vapor/console-kit.git resolved at 4.2.7
Creating working copy for https://github.com/vapor/websocket-kit.git
Working copy of https://github.com/vapor/websocket-kit.git resolved at 2.3.0
Creating working copy for https://github.com/vapor/routing-kit.git
Working copy of https://github.com/vapor/routing-kit.git resolved at 4.3.1
Creating working copy for https://github.com/apple/swift-nio-extras.git
Working copy of https://github.com/apple/swift-nio-extras.git resolved at 1.10.2
Creating working copy for https://github.com/swift-server/async-http-client.git
Working copy of https://github.com/swift-server/async-http-client.git resolved at 1.9.0
Creating working copy for https://github.com/apple/swift-crypto.git
Working copy of https://github.com/apple/swift-crypto.git resolved at 2.0.5
Creating working copy for https://github.com/apple/swift-metrics.git
Working copy of https://github.com/apple/swift-metrics.git resolved at 2.3.0
Creating working copy for https://github.com/swift-server/swift-backtrace.git
Working copy of https://github.com/swift-server/swift-backtrace.git resolved at 1.3.1
Creating working copy for https://github.com/apple/swift-nio-transport-services.git
Working copy of https://github.com/apple/swift-nio-transport-services.git resolved at 1.11.4
Creating working copy for https://github.com/vapor/multipart-kit.git
Working copy of https://github.com/vapor/multipart-kit.git resolved at 4.5.1
Creating working copy for https://github.com/apple/swift-nio-ssl.git
Working copy of https://github.com/apple/swift-nio-ssl.git resolved at 2.18.0
Creating working copy for https://github.com/apple/swift-nio-http2.git
Working copy of https://github.com/apple/swift-nio-http2.git resolved at 1.20.1
Creating working copy for https://github.com/vapor/async-kit.git
Working copy of https://github.com/vapor/async-kit.git resolved at 1.11.1
Creating working copy for https://github.com/apple/swift-nio.git
Working copy of https://github.com/apple/swift-nio.git resolved at 2.39.0
Creating working copy for https://github.com/apple/swift-log.git
Working copy of https://github.com/apple/swift-log.git resolved at 1.4.2
{
  "identity": "vapor",
  "name": "vapor",
  "url": "/Users/marcel/workspace/playground/vapor",
  "version": "unspecified",
  "path": "/Users/marcel/workspace/playground/vapor",
  "dependencies": [
    {
      "identity": "async-http-client",
      "name": "async-http-client",
      "url": "https://github.com/swift-server/async-http-client.git",
      "version": "1.9.0",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/async-http-client",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        },
        {
          "identity": "swift-nio-ssl",
          "name": "swift-nio-ssl",
          "url": "https://github.com/apple/swift-nio-ssl.git",
          "version": "2.18.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-ssl",
          "dependencies": [
            {
              "identity": "swift-nio",
              "name": "swift-nio",
              "url": "https://github.com/apple/swift-nio.git",
              "version": "2.39.0",
              "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
              "dependencies": [

              ]
            }
          ]
        },
        {
          "identity": "swift-nio-http2",
          "name": "swift-nio-http2",
          "url": "https://github.com/apple/swift-nio-http2.git",
          "version": "1.20.1",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-http2",
          "dependencies": [
            {
              "identity": "swift-nio",
              "name": "swift-nio",
              "url": "https://github.com/apple/swift-nio.git",
              "version": "2.39.0",
              "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
              "dependencies": [

              ]
            }
          ]
        },
        {
          "identity": "swift-nio-extras",
          "name": "swift-nio-extras",
          "url": "https://github.com/apple/swift-nio-extras.git",
          "version": "1.10.2",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-extras",
          "dependencies": [
            {
              "identity": "swift-nio",
              "name": "swift-nio",
              "url": "https://github.com/apple/swift-nio.git",
              "version": "2.39.0",
              "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
              "dependencies": [

              ]
            }
          ]
        },
        {
          "identity": "swift-nio-transport-services",
          "name": "swift-nio-transport-services",
          "url": "https://github.com/apple/swift-nio-transport-services.git",
          "version": "1.11.4",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-transport-services",
          "dependencies": [
            {
              "identity": "swift-nio",
              "name": "swift-nio",
              "url": "https://github.com/apple/swift-nio.git",
              "version": "2.39.0",
              "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
              "dependencies": [

              ]
            }
          ]
        },
        {
          "identity": "swift-log",
          "name": "swift-log",
          "url": "https://github.com/apple/swift-log.git",
          "version": "1.4.2",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-log",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "async-kit",
      "name": "async-kit",
      "url": "https://github.com/vapor/async-kit.git",
      "version": "1.11.1",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/async-kit",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        },
        {
          "identity": "swift-log",
          "name": "swift-log",
          "url": "https://github.com/apple/swift-log.git",
          "version": "1.4.2",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-log",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "console-kit",
      "name": "console-kit",
      "url": "https://github.com/vapor/console-kit.git",
      "version": "4.2.7",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/console-kit",
      "dependencies": [
        {
          "identity": "swift-log",
          "name": "swift-log",
          "url": "https://github.com/apple/swift-log.git",
          "version": "1.4.2",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-log",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "swift-crypto",
      "name": "swift-crypto",
      "url": "https://github.com/apple/swift-crypto.git",
      "version": "2.0.5",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-crypto",
      "dependencies": [

      ]
    },
    {
      "identity": "routing-kit",
      "name": "routing-kit",
      "url": "https://github.com/vapor/routing-kit.git",
      "version": "4.3.1",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/routing-kit",
      "dependencies": [

      ]
    },
    {
      "identity": "swift-backtrace",
      "name": "swift-backtrace",
      "url": "https://github.com/swift-server/swift-backtrace.git",
      "version": "1.3.1",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-backtrace",
      "dependencies": [

      ]
    },
    {
      "identity": "swift-nio",
      "name": "swift-nio",
      "url": "https://github.com/apple/swift-nio.git",
      "version": "2.39.0",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
      "dependencies": [

      ]
    },
    {
      "identity": "swift-nio-ssl",
      "name": "swift-nio-ssl",
      "url": "https://github.com/apple/swift-nio-ssl.git",
      "version": "2.18.0",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-ssl",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "swift-nio-http2",
      "name": "swift-nio-http2",
      "url": "https://github.com/apple/swift-nio-http2.git",
      "version": "1.20.1",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-http2",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "swift-nio-extras",
      "name": "swift-nio-extras",
      "url": "https://github.com/apple/swift-nio-extras.git",
      "version": "1.10.2",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-extras",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        }
      ]
    },
    {
      "identity": "swift-log",
      "name": "swift-log",
      "url": "https://github.com/apple/swift-log.git",
      "version": "1.4.2",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-log",
      "dependencies": [

      ]
    },
    {
      "identity": "swift-metrics",
      "name": "swift-metrics",
      "url": "https://github.com/apple/swift-metrics.git",
      "version": "2.3.0",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-metrics",
      "dependencies": [

      ]
    },
    {
      "identity": "websocket-kit",
      "name": "websocket-kit",
      "url": "https://github.com/vapor/websocket-kit.git",
      "version": "2.3.0",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/websocket-kit",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        },
        {
          "identity": "swift-nio-ssl",
          "name": "swift-nio-ssl",
          "url": "https://github.com/apple/swift-nio-ssl.git",
          "version": "2.18.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio-ssl",
          "dependencies": [
            {
              "identity": "swift-nio",
              "name": "swift-nio",
              "url": "https://github.com/apple/swift-nio.git",
              "version": "2.39.0",
              "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
              "dependencies": [

              ]
            }
          ]
        }
      ]
    },
    {
      "identity": "multipart-kit",
      "name": "multipart-kit",
      "url": "https://github.com/vapor/multipart-kit.git",
      "version": "4.5.1",
      "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/multipart-kit",
      "dependencies": [
        {
          "identity": "swift-nio",
          "name": "swift-nio",
          "url": "https://github.com/apple/swift-nio.git",
          "version": "2.39.0",
          "path": "/Users/marcel/workspace/playground/vapor/.build/checkouts/swift-nio",
          "dependencies": [

          ]
        }
      ]
    }
  ]
}
```
</details>

To avoid this, you can write the dependency tree to a temporary file:

```
swift package show-dependencies --format json --output-path tmp.json
```

Reply:
> `swift package show-dependencies --format json` prints "debug" output also to `stdout` which then cannot be parsed as a JSON.

FWIW, at least Swift version 5.8.1 that I tested with prints the non-JSON output to stderr instead of stdout.